File: backends/java/zappy-application/src/main/java/com/zappymart/application/promotions/ApplyPromotionCode.java
package com.zappymart.application.promotions;

import com.zappymart.application.Visitor;
import com.zappymart.application.cart.CurrentCart;
import com.zappymart.application.ports.CartRepository;
import com.zappymart.application.ports.Clock;
import com.zappymart.application.ports.PromotionRepository;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.domain.cart.Cart;
import com.zappymart.domain.promotions.AppliedPromotion;
import com.zappymart.domain.promotions.Promotion;
import com.zappymart.domain.promotions.PromotionCode;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

import java.util.Optional;

public final class ApplyPromotionCode {

    private final UnitOfWork unitOfWork;
    private final CurrentCart currentCart;
    private final CartRepository cartRepository;
    private final PromotionRepository promotionRepository;
    private final Clock clock;

    public ApplyPromotionCode(UnitOfWork unitOfWork, CurrentCart currentCart, CartRepository cartRepository,
                              PromotionRepository promotionRepository, Clock clock) {
        this.unitOfWork = unitOfWork;
        this.currentCart = currentCart;
        this.cartRepository = cartRepository;
        this.promotionRepository = promotionRepository;
        this.clock = clock;
    }

    public Result<Cart> execute(Visitor visitor, String typedCode) {
        return unitOfWork.inTransaction(() -> {
            Optional<PromotionCode> code = PromotionCode.parse(typedCode);
            if (code.isEmpty()) {
                return unknownCode(typedCode);
            }
            Optional<Promotion> promotion = promotionRepository.byCode(code.get());
            if (promotion.isEmpty()) {
                return unknownCode(code.get().value());
            }
            Cart cart = currentCart.forVisitor(visitor);
            Result<AppliedPromotion> applied = promotion.get().applyTo(cart.subtotal(), clock.now());
            if (!applied.succeeded()) {
                return applied.carryRefusal();
            }
            Cart changed = cart.withPromotion(code.get(), promotion.get().rule(), clock.now());
            return Result.of(cartRepository.save(changed));
        });
    }

    private Result<Cart> unknownCode(String typedCode) {
        return Result.refuse(UserErrorCode.CODE_UNKNOWN, "No promotion code reads " + typedCode + ".", "code");
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/cart/ApplyPromotionCode.kt
package nl.zappymart.application.cart

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.CartRepository
import nl.zappymart.application.ports.UnitOfWork
import nl.zappymart.domain.shared.Result

class ApplyPromotionCode(
    private val visitorCart: VisitorCart,
    private val cartPromotion: CartPromotion,
    private val carts: CartRepository,
    private val unitOfWork: UnitOfWork,
) {

    fun execute(visitor: Visitor, code: String): CartChange = unitOfWork.execute {
        val cart = cartPromotion.refreshed(visitorCart.of(visitor))
        when (val changed = cartPromotion.applying(cart, code)) {
            is Result.Success -> CartChange.made(carts.save(changed.value))
            is Result.Refused -> CartChange.refused(cart, changed.errors, null)
        }
    }
}
